### PR TITLE
Update send/bring shortcuts and show them properly per operating system

### LINKED ISSUE #850: Bring/Send Forward/Backward Shortcuts don't work on macOS
None of them work for me using macOS / Chrome latest.

### PR DESCRIPTION
**Pull request by lipis**
Changed the following (Like in Sketch, Figma):

- Bring Forward `Cmd + ]` 
- Send Backward `Cmd + [`
- Bring to Front `Option + Cmd + ]`
- Sent to Back `Option + Cmd + [`

Test the ZoomIn/ZoomOut titles on different operating systems..

- [x] Mac
- [x] Windows
- [x] Linux

Screenshot 2020-02-20 at 9 55 07 PM

Fixes #850

### CODE REVIEW

**Comment by lipis** [comment]
Can someone confirm that the zoom in / zoom out shortcuts appear correct on other operating systems?

File: src/actions/actionZindex.tsx [hunk 1]
```diff
@@ -86,13 +87,13 @@ export const actionSendBackward: Action = {
   contextItemLabel: "labels.sendBackward",
   keyPriority: 40,
   commitToHistory: () => true,
-  keyTest: event => event[KEYS.META] && event.altKey && event.key === "B",
+  keyTest: event => event[KEYS.CTRL_OR_CMD] && event.key === "[",
```
Comment (lipis) [inline src/actions/actionZindex.tsx]: Does anyone know why this one is not working?

File: src/actions/actionZindex.tsx [hunk 2]
```diff
@@ -110,13 +111,13 @@ export const actionBringForward: Action = {
   contextItemLabel: "labels.bringForward",
   keyPriority: 40,
   commitToHistory: () => true,
-  keyTest: event => event[KEYS.META] && event.altKey && event.key === "F",
+  keyTest: event => event[KEYS.CTRL_OR_CMD] && event.key === "]",
```
Comment (lipis) [inline src/actions/actionZindex.tsx]: and this one..

File: src/actions/actionZindex.tsx (hunk 1 above)
Comment (dwelle) [reply src/actions/actionZindex.tsx]: Works for me. What browser are you using? Btw, is the `CMD + [` a standard here? I'm just asking, no idea if what we used before is or isn't.

Comment (lipis) [reply src/actions/actionZindex.tsx]: Chrome / Mac, all latest.. :/

File: src/actions/actionZindex.tsx [hunk 3]
```diff
@@ -133,13 +134,14 @@ export const actionSendToBack: Action = {
   },
   contextItemLabel: "labels.sendToBack",
   commitToHistory: () => true,
-  keyTest: event => event[KEYS.META] && event.shiftKey && event.key === "B",
+  keyTest: event =>
+    event[KEYS.CTRL_OR_CMD] && event.altKey && event.key === "[",
```
Comment (lipis) [inline src/actions/actionZindex.tsx]: My bad, this one is not working.

File: src/actions/actionZindex.tsx [hunk 4]
```diff
@@ -156,13 +158,14 @@ export const actionBringToFront: Action = {
   },
   commitToHistory: () => true,
   contextItemLabel: "labels.bringToFront",
-  keyTest: event => event[KEYS.META] && event.shiftKey && event.key === "F",
+  keyTest: event =>
+    event[KEYS.CTRL_OR_CMD] && event.altKey && event.key === "]",
```
Comment (lipis) [inline src/actions/actionZindex.tsx]: and this one..

File: src/actions/actionZindex.tsx (hunk 3 above)
Comment (dwelle) [reply src/actions/actionZindex.tsx]: Right, this is our priority system in work (or, more like the fact we have no priority system in place). It matches `event[KEYS.CTRL_OR_CMD] && event,key === "["` from above, and exits early. One short-term solution is to move this key above it so it matches first.

Comment (lipis) [reply src/actions/actionZindex.tsx]: Didn't work.. Feel free to commit directly into this branch if you know how to solve this.. I

Comment (dwelle) [reply src/actions/actionZindex.tsx]: I've just pushed what you've definitely tried, but for me it works :man_shrugging:

Comment (lipis) [reply src/actions/actionZindex.tsx]: `Cmd+Option+[`, `Cmd+Option+]`

Don't work for me.. Mac with Chrome, Safari with US layout keyboard... very strange

Comment (dwelle) [reply src/actions/actionZindex.tsx]: Mhm.. `Option` is still `event.altKey`, right?

Comment (dwelle) [reply src/actions/actionZindex.tsx]: I'm not sure what behavior you're seeing exactly. Is it matching the wrong hotkey, or not matching any at all? What `event.key` and modifiers are shown in the `index.tsx` `onKeyDown` handler?

Comment (lipis) [reply src/actions/actionZindex.tsx]: Just doesn't bring it to front nor back with that shortcut.. I'm guessing you are not on a Mac..

Comment (dwelle) [reply src/actions/actionZindex.tsx]: Nope, Windows. So it doesn't do anything... I'd debug the `KeyboardEvent` directly.

**Comment by j-f1** [comment]
Could it be that the browser overrides Cmd+[] as back/forward?

**Comment by lipis** [comment]
`Cmd+[]` works :/

File: src/actions/actionZindex.tsx (hunk 3 above)
Comment (lipis) [reply src/actions/actionZindex.tsx]: `Ctrl+Alt+[]` works for you on Windows?

Comment (dwelle) [reply src/actions/actionZindex.tsx]: Yep :/.

**Comment by Fausto95** [comment]
This doesn't work for non-querty keyboards

**Comment by lipis** [comment]
This doesn't work on my QWERY keyboard either :/ (mac)

**Comment by j-f1** [comment]
Maybe it should use `event.code === "BracketRight"`? I think the issue is that the default macOS keyboard layout aliases option+`[` `]` to `“` and `‘`, so `event.key` is `"‘"`. You could also check `event.keyIdentifier === 'U+005D'` or `e.keyCode = 221`.

**Comment by lipis** [comment]
I tried with all of them.. same result :) (Except the `keyIdentifier` TBH.. will try it later tonight)

**Comment by lipis** [comment]
The current behaviour is better than the existing one.. shall we merge and improve later?

**Comment by j-f1** [comment]
Could you do a check if `event.key === "[" || event.code === "BracketLeft"`?

**Comment by lipis** [comment]
> Could you do a check if `event.key === "[" || event.code === "BracketLeft"`?

nothing..

**Comment by j-f1** [comment]
This works for me in Safari:

```js
  keyTest: event =>
    event[KEYS.CTRL_OR_CMD] &&
    !event.altKey &&
    (event.key === "]" || event.code === "BracketRight"),
```

**Comment by dwelle** [comment]
Btw, I just checked with Figma and Affinity, and they both use `Ctrl + Shift + [/]` for send to back/forward. Not sure about Mac, but we should align with that.

**Comment by lipis** [comment]
It doesn't.. I was checking sketch for reference: https://www.sketchshortcuts.com/

**Comment by dwelle** [comment]
I'd still use Figma/Affinity shortcuts instead of Sketch-only.

**Comment by lipis** [comment]
yes.. go ahead.

**Comment by dwelle** [comment]
Can someone on Mac test the `Cmd+Shift+[/]`? https://excalidraw-2bme8zkpw.now.sh/

**Comment by lipis** [comment]
On Chrome it jumps between next/previous tabs

**Comment by dwelle** [comment]
That's weird. On Figma it works fine, and we're also preventing the default (AFAIK). Is it possibly related only to your case? Can you repro @j-f1?

**Comment by lipis** [comment]
Doesn't work on Safari either..

**Comment by dwelle** [comment]
So here's the deal and my latest understanding:

- If we can stick to the industry standard shortcuts that this PR introduced (`Cmd/Ctrl+Alt/Shift+[/]`), we can either use `event.key` so we always use `[/]` printable character, regardless of keyboard layout (but this will come at a cost that some keyboard layouts such as AZERTY won't work at all because they 1) use their own modifier combinations to type that character, which clashes with our `Cmd/Ctrl+Alt/Shift` modifiers, and 2) the `[/]` char is under a number key, which in combination with `Cmd` Safari considers a tab switch), or we can use `event.code` which is layout-agnostic, and is always mapped to the same physical key (placement-wise, regardless of printable character it produces). Problem is, on many layouts this doesn't correspond to `[/]` label at all, so if we display a hint of `Cmd+Shift+[`, AZERTY users won't have a clue which actual key to press (on Mac AZERTY, that would be `^`) --- AFAIK there's no way to dynamically retrieve a printable key name from an `event.keyCode` (unless we use `keypress` event, but in that case that wouldn't work either because a key press that uses a modifier key won't trigger the event in the first place).
- We can go back to the non-standard `Cmd+Option+F/B`, which is more layout-compatible.

Note that even if we decide to revert the shortcut back to `Cmd+Option+F/B`, which will solve this specific case, it will not help us in similar cases for other shortcuts (current or future) --- mainly because there are many keyboard layouts out there and we can't easily predict which shortcuts will be affected (and even if we did, we won't be able to ensure every possible layout works).

**TLDR:** it's almost impossible to satisfy every keyboard layout, every platform, and every browser, at the same time. It's a mine field.

PS: it seems that even desktop apps (such as Affinity) don't handle AZERTY layouts correctly, nor do web apps (Figma). So we may go similar route and simply throw some users under the bus because we'll have to, eventually (if not already).

**Comment by lipis** [comment]
Let's merge this and improve later.. If most cases are working with `Cmd/Ctrl+Alt/Shift+[/]` we are already better...

**Comment by dwelle** [comment]
@lipis I'll fix the conflicts unless you're already on it.

**Comment by lipis** [comment]
no.. go ahead.. fix and merge

File: src/components/App.tsx [hunk 5]
```diff
@@ -1769,7 +1769,7 @@ export class App extends React.Component<any, AppState> {
     event.preventDefault();
     const { deltaX, deltaY } = event;
 
-    if (event.metaKey || event.ctrlKey) {
```
Comment (dwelle) [inline src/components/App.tsx]: @j-f1 any reason you moved away from platform-specific key to a union?

Comment (dwelle) [reply src/components/App.tsx]: I'll merge this. If this change was intentional, we can fix this later.

Comment (j-f1) [reply src/components/App.tsx]: Yes. Ctrl+scroll is what’s emitted by Chrome and Firefox when the user pinches on the trackpad.

Comment (dwelle) [reply src/components/App.tsx]: Pinch is handled by `wheel` event?

Comment (j-f1) [reply src/components/App.tsx]: Yes. Safari uses nonstandard `gesture*` events, but the other browsers use a `wheel` event.

Comment (dwelle) [reply src/components/App.tsx]: I see. Will make a fix.
